pilot/pkg/proxy/envoy/v2/xds.go

```diff
@@ -24,75 +27,135 @@ import (
 	"google.golang.org/grpc"
 
 	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/v1alpha3"
 )
 
 // ConfigCache for xDS resources
 type ConfigCache struct {
+	serviceSuffix string
+
+	// inputs
 	services model.ServiceDiscovery
 	configs  model.ConfigStore
 
+	// version counter
+	// TODO: this should be derived from the inputs for load balancing between xDS instances to work correctly
+	version int64
+
+	// generated hosts for the version counter;
+	// hosts need to be computed once per the entire mesh, reducing the load on specializing it to each proxy
+	hosts []v1alpha3.GuardedHost
```

Comment:
why can't you simply compute the entire route configuration? You cannot share the routes across namespaces if thats what you are thinking, because the clusters are generated based on the FQDN, which has the namespace encoded. 

This is also needless kubernetes complexity spilling into our code. We are far better off asking people to always use full names instead of annoying short names with ambiguous interpretation semantics.

Reply:
(cc @louiscryan) as we discussed multiple times, services do not have the same short names across namespaces. Service "reviews.default.svc.cluster.local" can be addressed as "reviews" from a proxy in "default", but not from a proxy in "test" namespace. This is not an emotional problem (annoying?), it's a correctness issue. Every other k8s app should respect this convention.